File: libraries/fary/ary1_ffs.f
      SUBROUTINE ARY1_FFS( BLOCK, SLOT, STATUS )
*+
*  Name:
*     ARY1_FFS

*  Purpose:
*     Find a free slot in a common block.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY1_FFS( BLOCK, SLOT, STATUS )

*  Description:
*     The routine finds a free slot in one of the common blocks
*     maintained by the ARY_ system. The number of a free slot is
*     returned and the slot is marked as used. An error is reported if
*     no further free slots exist.

*  Arguments:
*     BLOCK = INTEGER (Given)
*        The block in which a free slot is required.  The integer
*        symbolic constants ARY__ACB, ARY__DCB, ARY__MCB and ARY__PCB
*        are available to identify these.
*     SLOT = INTEGER (Returned)
*        The number of a free slot in the requested block.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Notes:
*     -  If STATUS is set on entry, then a SLOT value of zero will be
*     returned, although no further processing will occur.
*     -  A SLOT value of zero will also be returned if the routine
*     should fail for any reason.

*  Algorithm:
*     -  Set an initial value of SLOT before checking the inherited
*     status.
*     -  The relevant block is searched to find a slot which has not
*     been marked as used.
*     -  If a free slot is found, it is marked as used and its number is
*     returned.
*     -  If there is no free slot available, then report an error.
*     -  Report an error if the BLOCK argument supplied is not valid.

*  Copyright:
*     Copyright (C) 1989 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*     
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*     
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place,Suite 330, Boston, MA
*     02111-1307, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     22-MAY-1989 (RFWS):
*        Original version.
*     12-JUN-1989 (RFWS):
*        Adapted to new common block structure.
*     16-AUG-1989 (RFWS):
*        Changed so that SLOT is always returned as zero unless the
*        routine succeeds.
*     18-AUG-1989 (RFWS):
*        Changed to use a message token for the routine name, to prevent
*        '$' from affecting the message.
*     19-SEP-1989 (RFWS):
*        Added support for the Placeholder Control Block.
*     9-OCT-1989 (RFWS):
*        Updated prologue description of the BLOCK argument.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_PAR'          ! ARY_ public constants
      INCLUDE 'ARY_CONST'        ! ARY_ private constants
      INCLUDE 'ARY_ERR'          ! ARY_ error codes

*  Global Variables:
      INCLUDE 'ARY_DCB'          ! ARY_ Data Control Block
*        DCB_USED = LOGICAL( ARY__MXDCB ) (Read and Write)
*           Whether a slot in the DCB has been used.

      INCLUDE 'ARY_ACB'          ! ARY_ Access Control Block
*        ACB_USED = LOGICAL( ARY__MXACB ) (Read and Write)
*           Whether a slot in the ACB has been used.

      INCLUDE 'ARY_MCB'          ! ARY_ Mapping Control Block
*        MCB_USED = LOGICAL( ARY__MXMCB ) (Read and Write)
*           Whether a slot in the MCB has been used.

      INCLUDE 'ARY_PCB'          ! ARY_ Placeholder Control Block
*        PCB_USED = LOGICAL( ARY__MXPCB ) (Read and Write)
*           Whether a slot in the PCB has been used.

*  Arguments Given:
      INTEGER BLOCK

*  Arguments Returned:
      INTEGER SLOT

*  Status:
      INTEGER STATUS             ! Global status

*  External references:
      EXTERNAL ARY1_INIT         ! Initialise common blocks

*  Local variables:
      INTEGER I                  ! Loop counter for slots

*.

*  Set an initial value for the SLOT argument.
      SLOT = 0

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  If requested, search for a free slot in the Data Control Block.
      IF ( BLOCK .EQ. ARY__DCB ) THEN
         DO 1 I = 1, ARY__MXDCB
            IF ( .NOT. DCB_USED( I ) ) THEN
               DCB_USED( I ) = .TRUE.
               SLOT = I
               GO TO 2
            END IF
1        CONTINUE

*  If no free slot could be found, then report an error.
         STATUS = ARY__EXSLT
         CALL MSG_SETI( 'NSLOT', ARY__MXDCB )
         CALL ERR_REP( 'ARY1_FFS_DCB',
     :   'All ^NSLOT slots allocated for entries in the ARY_ ' //
     :   'facility Data Control Block have been used up.', STATUS )
2        CONTINUE

*  If requested, search for a free slot in the Access Control Block.
      ELSE IF ( BLOCK .EQ. ARY__ACB ) THEN
         DO 3 I = 1, ARY__MXACB
            IF ( .NOT. ACB_USED( I ) ) THEN
               ACB_USED( I ) = .TRUE.
               SLOT = I
               GO TO 4
            END IF
3        CONTINUE

*  If no free slot could be found, then report an error.
         STATUS = ARY__EXSLT
         CALL MSG_SETI( 'NSLOT', ARY__MXACB )
         CALL ERR_REP( 'ARY1_FFS_ACB',
     :   'All ^NSLOT slots allocated for entries in the ARY_ ' //
     :   'facility Access Control Block have been used up.', STATUS )
4        CONTINUE

*  If requested, search for a free slot in the Mapping Control Block.
      ELSE IF ( BLOCK .EQ. ARY__MCB ) THEN
         DO 5 I = 1, ARY__MXMCB
            IF ( .NOT. MCB_USED( I ) ) THEN
               MCB_USED( I ) = .TRUE.
               SLOT = I
               GO TO 6
            END IF
5        CONTINUE

*  If no free slot could be found, then report an error.
         STATUS = ARY__EXSLT
         CALL MSG_SETI( 'NSLOT', ARY__MXMCB )
         CALL ERR_REP( 'ARY1_FFS_MCB',
     :   'All ^NSLOT slots allocated for entries in the ARY_ ' //
     :   'facility Mapping Control Block have been used up.', STATUS )
6        CONTINUE

*  If requested, search for a free slot in the Placeholder Control
*  Block.
      ELSE IF ( BLOCK .EQ. ARY__PCB ) THEN
         DO 7 I = 1, ARY__MXPCB
            IF ( .NOT. PCB_USED( I ) ) THEN
               PCB_USED( I ) = .TRUE.
               SLOT = I
               GO TO 8
            END IF
7        CONTINUE

*  If no free slot could be found, then report an error.
         STATUS = ARY__EXSLT
         CALL MSG_SETI( 'NSLOT', ARY__MXPCB )
         CALL ERR_REP( 'ARY1_FFS_PCB',
     :   'All ^NSLOT slots allocated for entries in the ARY_ ' //
     :   'facility Placeholder Control Block have been used up.',
     :   STATUS )
8        CONTINUE

*  If the block specified was invalid, then report an error.
      ELSE
         STATUS = ARY__FATIN
         CALL MSG_SETC( 'ROUTINE', 'ARY1_FFS' )
         CALL MSG_SETI( 'BADBLOCK', BLOCK )
         CALL ERR_REP( 'ARY1_FFS_BLOCK',
     :   'Routine ^ROUTINE called with an invalid BLOCK argument of ' //
     :   '^BADBLOCK (internal programming error).', STATUS )
      END IF

*  Call error tracing routine and exit.
      IF( STATUS .NE. SAI__OK ) CALL ARY1_TRACE( 'ARY1_FFS', STATUS )

      END

File: libraries/ary/ary1Ffs.c
#include <pthread.h>
#include "sae_par.h"
#include "ary1.h"
#include "mers.h"
#include "ary_err.h"
#include "ary_ast.h"

AryDCB **Ary_DCB = NULL;  /* Pointer to array of all DCB pointers */
AryACB **Ary_ACB = NULL;  /* Pointer to array of all ACB pointers */
AryMCB **Ary_MCB = NULL;  /* Pointer to array of all MCB pointers */
AryPCB **Ary_PCB = NULL;  /* Pointer to array of all PCB pointers */

int Ary_NDCB = 0;    /* Number of DCBs in above array */
int Ary_NACB = 0;    /* Number of ACBs in above array */
int Ary_NMCB = 0;    /* Number of MCBs in above array */
int Ary_NPCB = 0;    /* Number of PCBs in above array */

/* A mutex to serialise access to each of the above array. A thread
   should only access an array after it has acquired a lock on this mutex. */
pthread_mutex_t Ary_DCB_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t Ary_ACB_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t Ary_MCB_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t Ary_PCB_mutex = PTHREAD_MUTEX_INITIALIZER;

void *ary1Ffs( AryBlockType type, int *status ) {
/*
*+
*  Name:
*     ary1Ffs

*  Purpose:
*     Allocate an ARY "block" structure of any type.

*  Synopsis:
*     void *ary1Ffs( AryBlockType type, int *status )

*  Description:
*     The routine finds a free slot in one of the arrays holding pointers
*     to "block" structures within by the ARY_ system. The pointer stored
*     in the free slot is returned and the structure is marked as used.
*     A new structure is created and the array of pointers is extended
*     if no further free slots exist.
*
*     An array of pointers is used, with each actual structure being
*     allocated separately, so that extending the array using realloc
*     does not change the pointers to the structures. If an array of
*     structures had been used, extending it using realloc would have
*     changed the address of each structure.

*  Parameters:
*     type
*        The type of "block" structure to be allocated. The constants
*        ARY__ACBTYPE, ARY__DCBTYPE, ARY__MCBTYPE and ARY__PCBTYPE are
*        available to identify these.
*     status
*        The global status.

*  Returned function value:
*     A pointer to the object held in the allocated slot of the array.
*     The returned pointer should be cast to the required type (AryDCB,
*     AryACB, etc).

*  Notes:
*     -  The returned pointer should be released using ary1Rls when it is
*     no longer needed.
*     -  If STATUS is set on entry, then a NULL value will be returned,
*     although no further processing will occur.
*     -  A NULL value will also be returned if the routine should fail
*     for any reason.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     DSB: David S. Berry (EAO)

*  History:
*     28-JUL-2017 (DSB):
*        Original version.

*-
*/

/* Local variables: */
   AryObject **parray;        /* Pointer to start of pointer array */
   AryObject **head;          /* Pointer to next pointer */
   AryObject *result;         /* The returned pointer */
   const char *name;          /* Pointer to name string */
   int *pn;                   /* Pointer to variable hold size of array */
   int i;                     /* Loop counter for slots */
   int oldsize;               /* Original size of array */
   pthread_mutex_t *mutex;    /* Pointer to mutex for selected array */
   size_t size;               /* Size of each structure in array */

/* Set an initial value for the returned pointer. */
   result = NULL;

/* Check inherited global status. */
   if( *status != SAI__OK ) return NULL;

/* Store info about the required type of block. */
   if( type == ARY__DCBTYPE ){

/* A mutex to serialise access to the array of structure pointers */
      mutex = &Ary_DCB_mutex;

/* A pointer to the first structure in the array of allocated structures
   of the required type. */
      parray = (AryObject **) Ary_DCB;

/* Pointer to variable holding the current length of the array. */
      pn = &Ary_NDCB;

/* The size of each structure */
      size = sizeof(AryDCB);

/* Name string for error messages. */
      name = "DCB";

   } else if( type == ARY__ACBTYPE ){
      mutex = &Ary_ACB_mutex;
      parray = (AryObject **) Ary_ACB;
      pn = &Ary_NACB;
      size = sizeof(AryACB);
      name = "ACB";

   } else if( type == ARY__MCBTYPE ){
      mutex = &Ary_MCB_mutex;
      parray = (AryObject **) Ary_MCB;
      pn = &Ary_NMCB;
      size = sizeof(AryMCB);
      name = "MCB";

   } else if( type == ARY__PCBTYPE ){
      mutex = &Ary_PCB_mutex;
      parray = (AryObject **) Ary_PCB;
      pn = &Ary_NPCB;
      size = sizeof(AryPCB);
      name = "PCB";

/* If the block type specified was invalid, then report an error. */
   } else {
      *status = ARY__FATIN;
      msgSetc( "ROUTINE", "ary1Ffs" );
      msgSeti( "BADTYPE", type );
      errRep( " ", "Routine ^ROUTINE called with an invalid 'type' argument of "
              "^BADBLOCK (internal programming error).", status );
   }

/* Only proceed if the block type was recognised. */
   if( *status == SAI__OK ){

/* Wait until the current thread has exclusive access to the array. */
      pthread_mutex_lock( mutex );

/* Loop through the array looking for an element that is not currently in
   use. If found, use it as the returned result and indicate it is now in
   use. */
      head = parray;
      for( i = 0; i < *pn; i++,head++ ){
         if( !(*head)->used ){
            result = *head;
            result->used = 1;
            break;
         }
      }

/* If no free slot could be found, extend the array of DCB pointers by 100%,
   enforcing an initial size of 100. */
      if( !result ) {
         oldsize = *pn;
         if( oldsize ) {
            *pn *= 2;
         } else {
            *pn = 100;
         }

         parray = astGrow( parray, *pn, sizeof(*parray) );

/* Store the new pointer to the re-allocated array back in the correct
   global variable. */
         if( type == ARY__DCBTYPE ){
            Ary_DCB = (AryDCB **) parray;
         } else if( type == ARY__ACBTYPE ){
            Ary_ACB = (AryACB **) parray;
         } else if( type == ARY__MCBTYPE ){
            Ary_MCB = (AryMCB **) parray;
         } else if( type == ARY__PCBTYPE ){
            Ary_PCB = (AryPCB **) parray;
         }

/* Report an error if the array could not be extended. */
         if( *status != SAI__OK ) {
            errRepf( "", "ary1Ffs: Failed to extend array of %s objects.",
                    status, name );

/* If the array of pointers was extended successfully, populate the new
   elements using pointers to newly allocated DCB structures. Store the
   index of the structure in its "slot" component, and indicate it is not
   currently in use. */
         } else {
            head = parray + oldsize;
            for( i = oldsize; i < *pn; i++ ) {
               result = *(head++) = astCalloc( 1, size );
               if( *status == SAI__OK ) {
                  result->used = 0;
                  result->slot = i;
                  result->type = type;
               } else {
                  break;
               }
            }

/* If the new structures were created successfully, return the first of
   the new elements. */
            if( *status == SAI__OK ) {
               result = parray[ oldsize ];
               result->used = 1;
            }
         }
      }
   }

/* Release the lock on the array, this allowing any waiting threads
   to proceed. */
   pthread_mutex_unlock( mutex );

/* Call error tracing routine and exit. */
   if( *status != SAI__OK ) ary1Trace( "ary1Ffs", status );

   return (void *) result;
}
